### PR TITLE
VReplication: use proper column collations in vstreamer

### PR DESCRIPTION
**Pull request by mattlord**
## Description

In order to address the problem described in https://github.com/vitessio/vitess/issues/15338, this PR does the following:
  - Leverages optional metadata in row based binary log's `TableMapEvent`s when present
  - Leverages the column collation information provided by the target mysqld instance when possible

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/15338

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/vstreamer/testenv/testenv.go [hunk 1]
```diff
@@ -41,7 +42,12 @@ import (
 	vttestpb "vitess.io/vitess/go/vt/proto/vttest"
 )
 
-const DBName = "vttest"
+const (
+	DBName               = "vttest"
+	DefaultCollationName = "utf8mb4_0900_ai_ci"
+)
+
+var DefaultCollationID = collations.MySQL8().LookupByName(DefaultCollationName)
```
Comment (dbussink) [inline go/vt/vttablet/tabletserver/vstreamer/testenv/testenv.go:50]: Why not `collations.MySQL8().DefaultConnectionCharset()`? That returns the exact same thing. You can then also lookup the name for that. I think that's more semantically correct and avoids duplicating the `utf8mb4_0900_ai_ci` as a default value.

File: go/mysql/binlog_event.go [hunk 2]
```diff
@@ -216,6 +216,13 @@ type TableMap struct {
 	// - If the metadata is one byte, only the lower 8 bits are used.
 	// - If the metadata is two bytes, all 16 bits are used.
 	Metadata []uint16
+
+	// ColumnCollationIDs contains information about the inherited
+	// or implied column default collation and any explicit per-column
+	// override for character based columns ONLY. This means that the
+	// array position needs to be mapped to the ordered list of
+	// character based columns in the table.
+	ColumnCollationIDs []uint64
```
Comment (dbussink) [inline go/mysql/binlog_event.go:225]: Should this be `[]collations.ID` instead?

Comment (mattlord) [reply go/mysql/binlog_event.go:225]: Yeah, definitely. I just hadn't cared much about that stuff just yet. No reason not to do it at this point though.

File: go/vt/vttablet/tabletserver/vstreamer/testenv/testenv.go (hunk 1 above)
Comment (mattlord) [reply go/vt/vttablet/tabletserver/vstreamer/testenv/testenv.go:50]: I did think about that as well and started off using it. I may go back to it -- don't feel strongly at all. It's an inconsequential part of this PR IMO and I'm fine either way.

File: go/mysql/binlog_event.go [hunk 3]
```diff
@@ -216,6 +218,13 @@ type TableMap struct {
 	// - If the metadata is one byte, only the lower 8 bits are used.
 	// - If the metadata is two bytes, all 16 bits are used.
 	Metadata []uint16
+
+	// ColumnCollationIDs contains information about the inherited
+	// or implied column default collation and any explicit per-column
+	// override for text based columns ONLY. This means that the
+	// array position needs to be mapped to the ordered list of
+	// text based columns in the table.
+	ColumnCollationIDs []collations.ID
```
Comment (vmg) [inline go/mysql/binlog_event.go:227]: I think a lot of the code is going to be easier to write and easier to follow if this slice has one entry for each column and you simply fill the non-text columns with a Binary collation.

**Review by vmg — APPROVED** [review]
Looks correct to me, although as I pointed out, storing only textual collations makes this needlessly complex.

File: go/mysql/binlog_event.go (hunk 3 above)
Comment (mattlord) [reply go/mysql/binlog_event.go:227]: Soooo... I agree! But unfortunately I'd have to perform the same logic to fill in those gaps as the binlog row metadata does not include the column's index or name. I'll think about it some more though.

File: go/mysql/binlog_event_rbr.go [hunk 4]
```diff
@@ -88,11 +107,20 @@ func (ev binlogEvent) TableMap(f BinlogFormat) (*TableMap, error) {
...
+		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "unexpected metadata end: got %v was expecting %v (data=%v)", pos, expectedEnd, data)
 	}
 
 	// A bit array that says if each column can be NULL.
-	result.CanBeNull, _ = newBitmap(data, pos, int(columnCount))
+	result.CanBeNull, read = newBitmap(data, pos, int(columnCount))
+	pos = read
+
+	// Read any text based column collation values provided in the optional metadata.
+	//log.Errorf("DEBUG: Remaining optional metadata bytes for %s: %v", result.Name, data[pos:])
```
Comment (mattlord) [inline go/mysql/binlog_event_rbr.go:118]: All of these DEBUG logging statements will be removed before merging.

File: go/mysql/binlog_event.go (hunk 3 above)
Comment (mattlord) [reply go/mysql/binlog_event.go:227]: I explored this path, but we don't have access to the collation ENV in order to properly fill in the gaps. So leaving it as-is for now, where users of this (vstreamer) have access to both the collation ENV and mysqld.

**Review by rohit-nayak-ps — APPROVED** [review]
Nice work ❤️

**Pull request merged**